zp-relayer/endpoints.ts

```diff
@@ -108,22 +109,82 @@ async function getTransactionsV2(req: Request, res: Response, next: NextFunction
 }
 
 async function getJob(req: Request, res: Response) {
+  enum JobStatus {
+    WAITING = 'waiting',
+    FAILED = 'failed',
+    SENDED = 'sended',
+    REVERTED = 'reverted',
+    COMPLETED = 'completed',
+  }
+
+  interface GetJobResponse {
+    createdOn: number
+    failedReason: null | string
+    finishedOn: null | number
+    state: JobStatus
+    txHash: null | string
+  }
+
   const jobId = req.params.id
-  const job = await poolTxQueue.getJob(jobId)
-  if (job) {
-    const state = await job.getState()
-    const txHash = job.returnvalue
-    const failedReason = job.failedReason
-    const createdOn = job.timestamp
-    const finishedOn = job.finishedOn
-
-    res.json({
-      state,
-      txHash,
-      failedReason,
-      createdOn,
-      finishedOn,
-    })
+
+  async function getPoolJobState(requestedJobId: string): Promise<GetJobResponse | null> {
+    const jobId = await pool.state.jobIdsMapping.get(requestedJobId)
+    const job = await poolTxQueue.getJob(jobId)
+    if (!job) return null
+
+    // Default result object
+    let result: GetJobResponse = {
+      createdOn: job.timestamp,
+      failedReason: null,
+      finishedOn: null,
+      state: JobStatus.WAITING,
+      txHash: null,
+    }
+
+    const poolJobState = await job.getState()
+    if (poolJobState === 'completed') {
+      // Transaction was included in optimistic state, waiting to be mined
+      const sentJobId = job.returnvalue[0][1]
+      const sentJob = await sentTxQueue.getJob(sentJobId)
+      // Should not happen here, but need to verify to be sure
+      if (!sentJob) throw new Error('Sent job not found')
+
+      const sentJobState = await sentJob.getState()
+      if (sentJobState === 'waiting' || sentJobState === 'active' || sentJobState === 'delayed') {
+        // Transaction is in re-send loop
+        const txHash = sentJob.data.prevAttempts.at(-1)?.[0]
+        result.state = JobStatus.SENDED
+        result.txHash = txHash || null
+      } else if (sentJobState === 'completed') {
+        const [txState, txHash] = sentJob.returnvalue
```

Comment:
We shouldn't see 'SKIPPED' state here, because such jobs are added to the jobs remapping hset, right?

Reply:
Yes, if a job was re-scheduled, it will be in mapping hset, and we will get actual mapped job here
https://github.com/zkBob/zeropool-relayer/blob/369eeb62c4d9dfff262a0c24b78223eb55e7dd73/zp-relayer/endpoints.ts#L131